spec.md

```diff
@@ -412,6 +412,16 @@ service GroupController {
     }
 }
 
+service SnapshotMetadata {
```

Comment:
i'm wondering about "stream" for the responses. we don't stream any other type of response. i notice that each of the response types already has `repeated` block metadata. in practice, how large do we expect these metadata chains to be? it also looks like there's a mechanism to resume from some offset (perhaps in case the stream is interrupted?) - which means that clients already have to keep track of what they've consumed.

why not just send a very large batch of metadata for each request? the golang default gRPC max message size is 4MiB. is that not good enough?

is the use of streaming an example of premature optimization?

Reply:
This is definitely a new gRPC pattern for the CSI API, but I don't think this is a case of premature optimization.

The amount of changed block data would be proportional to the size of the volume, the amount of change, and the mechanism chosen to represent changed areas (FIXED_LENGH or VARIABLE_LENGTH).  For example, assuming a 1TiB volume (2^40) and 1MiB blocks, there would be 1MiB such blocks; with 8KiB blocks there would 128MiB such blocks. In the worst case for FIXED_LENGTH all the blocks could have changed. In the worst case for VARIABLE_LENGTH every alternative block could have changed.  16 bytes (+ some overhead?) are used for each changed block datum, which should also be factored in.

Beyond the length of a single stream response, assuming a fixed slice length forces the client to support this size, which could break some memory conservative clients. (This is why there is an option for the client to request a `max_results` limit).

Recall too, that an earlier iteration of this KEP was criticized for passing all this metadata through the Kubernetes API server. The direct connection from the backup app to this new CSI service (sidecar) bypasses the K8s API server but cannot reduce the quantity of metadata!

Reply:
@jdef We discussed this in the SIG Data Protection WG meeting. Some relevant observations:

- We're dealing with metadata access here, and not a control path. There is no existing precedent for such operations in the current CSI spec.
- The performance of a gRPC stream would better than the comparative use of a classic pagination API.
   - A classic pagination API would involve multiple round-trip calls, versus a single call with the stream API.
   - A classic pagination API requires the SP to fill the next page of metadata to return. Depending on the underlying APIs, this may require the SP to reissue the original allocated/diff call and skip the metadata records already returned in previous calls.  In contrast, an SP server processing a stream API can send multiple responses without reissuing the allocated/diff call.

Reply:
the last bullet point is probably the most convincing here